Language/Haskell/GhcMod/Monad/Types.hs

```diff
@@ -129,10 +138,11 @@ data GhcModState = GhcModState {
       gmGhcSession   :: !(Maybe GmGhcSession)
     , gmComponents   :: !(Map ChComponentName (GmComponent GMCResolved (Set ModulePath)))
     , gmCompilerMode :: !CompilerMode
+    , gmMMappedFiles :: Map TargetId (StringBuffer, UTCTime)
```

Comment:
Ok, I think I kind of understand the idea now. A couple points you might have missed though (or where I lack understanding maybe):

1. Imports don't generally honor mappings, and only way to make this happen is load all mapped files as targets, which doesn't sound that great. I don't think this is a *huge* problem, since, in general, it only makes sense to map a couple files that are actually being passed as targets, and not whole project.
2. Only command where ghc-mod accepts module name is `browse` I think, so for other commands `TargetFile` is looked up as `TargetFile`
3. `guessTarget` favors `TargetFile` greatly: if it can interpret a given `String` as a filename, it will.

So from where I stand, it seems like concerns would be that `browse` and `HomeModuleGraph` don't honor mappings.

One way to fix this would be to parse module name from source files when loading mappings and store those in `Map ModuleName ...`, and then get filenames from HMG when loading targets. But that doesn't strike me as the most straightforward option, since, in general, we already know filenames when loading mapped files.

Maybe a better option would be to add *both* `TargetFile` and `TargetModule` to `gmMMappedFIles` when loading those. This way, we save on an HMG trip at a cost of a little more memory, and that would work in most cases even if it's impossible to determine module name due to errors.

Does that sound reasonable to you?